### DISCUSSION TITLE
Launching a command and pasting its output back

### DISCUSSION THREAD

**Discussion by max-baz** [kovidgoyal/kitty#5836] [Q&A] [answered]
Hello,

I want to achieve the behavior similar to `unicode_input` kitten, but with an arbitrary command:

- Press a hotkey in kitty
- A new overlay window is opened, running the command of my choice (predefined in kitty config)
- The command prints the output to stdout
- Overlay window is closed, and the output of that command is pasted in my original kitty window.

I tried quite a few combinations of `launch` and `combine`, but I can't just figure it out 😐 

I was able to achieve what I want by creating a "wrapper" kitten (you can see an example below), but it feels like I'm missing something super obvious, and the same should be achievable with some combo of `launch` commands... 

I'd be happy to hear any ideas, can I simplify this workflow? Can I avoid creating the wrapper kitten?

Thanks! 🙏 

`/usr/bin/emojis`:

```sh
#!/bin/sh
printf ':)\n:(' | fzf | tr -d '\n'
```

`~/.config/kitty/run.py`:

```python
#!/usr/bin/env python3

import subprocess
from typing import List

from kitty.boss import Boss


def main(args: List[str]) -> str:
    result = subprocess.run(args[1:], stdout=subprocess.PIPE)
    return result.stdout.decode("utf-8")


def handle_result(
    args: List[str], answer: str, target_window_id: int, boss: Boss
) -> None:
    w = boss.window_id_map.get(target_window_id)
    if w is not None:
        w.paste_text(answer)
```

`~/.config/kitty/kitty.conf`:

```conf
...
map kitty_mod+u kitten run.py /usr/bin/emojis
...
```

**Comment by page-down** [comment, marked as answer]
> ... can I simplify this workflow? Can I avoid creating the wrapper kitten?

You can also use remote control, in any case, you need to implement it in a script or kitten.

https://sw.kovidgoyal.net/kitty/remote-control/#kitty-send-text

```conf
map f1 launch --type overlay --allow-remote-control /path/to/script
```

@kovidgoyal
I noticed that `kitty @ send-text --stdin` does not work properly, the following command works fine in 0.26.5.

```shell
printf 'debug' | kitty @ send-text -m id:1 --stdin
```

The option `--match` seems to be missing the way to match the launching (parent) window of the overlay, regardless of whether the window is active, focused or not.

**Reply by max-baz** [reply]
Thanks! It does the job without having to create a wrapper kitten, and still allows me to keep all the kitty logic out of the script itself, which is therefore usable both as a kitty mapping, and as a script that could be executed even from another kind of terminal.

I just have a follow-up question, do you know if there's a way to get the text sent not as a simulation of keypresses, but as a "paste"? I don't know the proper term for this, what I mean is that in a text editor you can paste clipboard content as "one event", and it would be pasted as is, while if the clipboard content is being typed character by character, editor is trying to be smart, auto-indenting lines, auto-inserting brackets, etc., thus corrupting the paste.

One of the use-cases I have for this feature is exactly for clipboard manager - I press a shortcut in kitty, search through my clipboard history using fzf, press Enter and the selected entry is pasted back in my original terminal, which might be a text editor.

**Reply by page-down** [reply]
@maximbaz
> ... get the text sent ... as a "paste" ...

For now you can use the send-text remote control command to send bracketed-paste escape sequences.

**Comment by kovidgoyal** [comment]
On Thu, Dec 29, 2022 at 06:10:15PM -0800, page-down wrote:
> > ... can I simplify this workflow? Can I avoid creating the wrapper kitten?
> 
> You can also use remote control, in any case, you need to implement it in a script or kitten.
> 
> https://sw.kovidgoyal.net/kitty/remote-control/#kitty-send-text
> 
> ```conf
> map f1 launch --type overlay --allow-remote-control /path/to/script
> ```
> 
> @kovidgoyal
> I noticed that `kitty @ send-text --stdin` does not work properly, the following command works fine in 0.26.5.

Now fixed.

> 
> ```shell
> printf 'debug' | kitty @ send-text -m id:1 --stdin
> ```
> 
> The option `--match` seems to be missing the way to match the launching (parent) window of the overlay, regardless of whether the window is active, focused or not.

I dont think this is new, it has always been the case. Once can add
another matching criteria for it I guess.

**Reply by page-down** [reply]
@kovidgoyal

>> ... missing the way to match the launching (parent) window of the overlay ...
>
> ... add another matching criteria for it ...

Thanks, I see that it has been implemented in master.

However, I have an uncertainty before the next release, should `self` and `overlay_parent` be considered as values under `state`? Are these a `state`?

**Comment by kovidgoyal** [comment]
On Fri, Dec 30, 2022 at 04:24:00AM -0800, page-down wrote:
> @kovidgoyal
> 
> >> ... missing the way to match the launching (parent) window of the overlay ...
> >
> > ... add another matching criteria for it ...
> 
> Thanks, I see that it has been implemented in master.
> 
> However, I have an uncertainty before the next release, should `self` and `overlay_parent` be considered as values under `state`? Are these a `state`?

Yes they are state, you can move windows around thereby changing the
overlay parent. self is of course constant but it is still state.
